drivers/hx711/hx711.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2018 - 2020 Philipp-Alexander Blum <philipp-blum@jakiku.de>
+ *               2019 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_sensors
+ * @brief       Driver for the HX711 weigh scale ADC
+ *
+ * @{
+ * @file
+ * @brief       HX711 driver
+ *
+ * @author      Philipp-Alexander Blum <philipp-blum@jakiku.de>
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include "periph/gpio.h"
+#include "periph/gpio_util.h"
+#include "xtimer.h"
+
+#include "hx711.h"
+#include "hx711_params.h"
+
+/**
+ * @brief Needs to wait for device to be ready while initialization
+ */
+static int _hx711_is_ready(hx711_t *dev)
+{
+    return gpio_read(dev->params.dout) == 0;
+}
+
+static void _hx711_set_gain(hx711_t *dev, hx711_gain_t gain)
+{
+    dev->params.gain = gain;
+    gpio_clear(dev->params.sck);
+    hx711_read(dev);
+}
+
+static void _hx711_set_offset(hx711_t *dev, int32_t offset)
+{
+    dev->offset = offset;
+}
+
+void hx711_init(hx711_t *dev, const hx711_params_t *params)
+{
+    dev->params = *params;
+    dev->offset = 0;
+
+    gpio_init(params->sck, GPIO_OUT);
+    gpio_init(params->dout, GPIO_IN);
+
+    _hx711_set_gain(dev, params->gain);
+}
+
+int32_t hx711_read(hx711_t *dev)
+{
+    while (!_hx711_is_ready(dev)) {}
+
+    union {
+        int32_t i32;
+        struct {
+            uint8_t data[3];
+            uint8_t fill_byte;
+        } u8;
+    } value;
+
+    value.i32 = 0;
+
+    gpio_t sck = dev->params.sck;
+    gpio_t dout = dev->params.dout;
+
+    unsigned gain = dev->params.gain;
+
+    value.u8.data[2] = gpio_util_shiftin(dout, sck);
+    value.u8.data[1] = gpio_util_shiftin(dout, sck);
+    value.u8.data[0] = gpio_util_shiftin(dout, sck);
```

Comment:
That's a really good point and would lead to an issue. They are not enforced. I need to think about this. Do you have an idea?

Reply:
IMO, this issue could arise to all users of `gpio_util_shiftin()`, if LTO is enabled. So maybe the best is to extend it to also take a third argument specifying the maximum speed. This could be implemented using [`xtimer_periodic_wakeup()`](http://api.riot-os.org/group__sys__xtimer.html#ga77ff235d491f651378348a46519d9dba)

If the GPIO ABC API proposed in https://github.com/RIOT-OS/RIOT/pull/12015 would already be available, it would be even possible to provide a clock frequency argument (instead of a *maximum* clock frequency argument): With GPIO ABC it would be possible to match the clock frequency quite well. Maybe I should invest some time to polish that API and add "killer applications" on top of it to motivate the API.

So for now: The `xtimer_periodic_wakeup()` is the best option IMO. But I don't think that an API change will make it into the upcoming release, as this is not something usually merged during the feature freeze. But it is not too long until the next development cycle starts and PRs can of course be opened and reviewed at any point in time, only the merging is limited during feature freeze.